spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,157 @@
+    `old_name='random_state'` either by keyword or as a positional argument at
+    `position_name=1`. At the time of application, the name of the argument in the
+    function signature is manually changed to the new name, `rng`. If positional
+    use was allowed before, this is not changed.*
+
+    - If the function is called with both `random_state` and `rng`, the decorator
+      raises an error.
+    - If `random_state` is provided as a keyword argument, the decorator passes
+      `random_state` to the function's `rng` argument as a keyword, but the decorator
+      will emit a `DeprecationWarning` about the deprecation of keyword `random_state`.
+    - If `random_state` is provided as a positional argument, the decorator passes
+      `random_state` to the function's `rng` argument by position, but the decorator
+      will emit a `FutureWarning` about the changing interpretation of the argument.
+    - If `rng` is provided as a keyword argument, the decorator validates `rng` using
+      `numpy.random.default_rng` before passing it to the function.
+    - If neither `random_state` nor `rng` is provided, the decorator checks whether
+      `np.random.seed` has been used to set the global seed. If so, it emits a
+      `FutureWarning`, noting that usage of `numpy.random.seed` will eventually have
+      no effect. Either way, the decorator calls the function without explicitly
+      passing the `rng` argument.
+
+    To avoid warnings, a user must pass `rng` as a keyword or pass a `Generator`
+    object or `None` by position.
+
+    After the deprecation period, the decorator can be removed, and the function
+    can simply validate the `rng` argument by calling `np.random.default_rng(rng)`.
+
+    * A `FutureWarning` is emitted when the PRNG argument is used by
+      position. It indicates that the "Hinsen principle" (same
+      code yielding different results in two versions of the software)
+      will be violated, unless positional use is deprecated. Specifically:
+
+      - If `None` is passed by position, the function will change from being
+        seeded to being unseeded.
+      - If an integer is passed by position, the random stream will change.
+
+      We suggest that projects consider deprecating positional use of
+      `random_state`/`rng` (i.e., change their function signatures to
+      ``def my_func(..., *, rng=None)``); that might not make sense
+      for all projects, so this SPEC does not make that
+      recommendation, neither does this decorator enforce it.
+
+    Parameters
+    ----------
+    old_name : str
+        The old name of the PRNG argument (e.g. `seed` or `random_state`).
+    position_num : int, optional
+        The (0-indexed) position of the old PRNG argument (if accepted by position).
+        Maintainers are welcome to eliminate this argument and use, for example,
+        `inspect`, if preferred.
+    dep_version : str, optional
```

Comment:
We could change this to just accept `end_version` for simplicity. This was originally modeled after a function that changed the name of an argument, and the deprecation warnings specified the version in which the deprecation began. We currently only mention the end version, so we could just accept that directly.

Reply:
That'd be a good simplification.

Reply:
Will do; LMK about https://github.com/scientific-python/specs/pull/180/files#r1684414274.

Reply:
We actually went the other direction here due to
https://github.com/scientific-python/specs/pull/180#discussion_r1705103987.